torch/_dynamo/variables/torch.py

```diff
@@ -1096,72 +1005,78 @@ def old_speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
             graph_checkpoint, checkpoint = tx.output.graph, tx.copy_graphstate()
 
             def speculate_branch(branch):
+                # NB: 0 is predicate
+                ix = 1 if branch else 2
                 try:
-                    # NB: 0 is predicate
-                    ix = 1 if branch else 2
-                    return old_speculate_subgraph(
-                        args[ix], operands, graph_checkpoint, checkpoint
+                    ret_val, ret_graph, ret_lifted_freevars = speculate_subgraph(
+                        tx, args[ix], operands, graph_checkpoint, checkpoint
                     )
-                except ArgsMismatchError as e:
-                    raise UserError(UserErrorType.DYNAMIC_CONTROL_FLOW, str(e))
-
-            (
-                true_r,
+                # Reraise because we want to suggest workarounds
+                except Unsupported as e:
+                    raise UserError(UserErrorType.DYNAMIC_CONTROL_FLOW, str(e)) from e
+
+                if not isinstance(ret_val, TensorVariable):
+                    raise UserError(
+                        UserErrorType.DYNAMIC_CONTROL_FLOW,
+                        "Expected branch out type to be a single tensor",
+                    )
+                return ret_val, ret_graph, ret_lifted_freevars
+
+            (true_r, true_graph, true_lifted_freevars) = speculate_branch(True)
+            true_nn_modules = tx.copy_graphstate().output.nn_modules
+
+            (false_r, false_graph, false_lifted_freevars) = speculate_branch(False)
+            false_nn_modules = tx.copy_graphstate().output.nn_modules
+
+            # TODO (tmanlaibaatar) deduplicate this later
```

Comment:
Could you file an issue about this so we can track it? Ideally we would link to the issue in this comment.